### PR TITLE
Fix posix version of moveFile between different filesystems

### PR DESCRIPTION
**Pull request by vegansk**
This PR fixes issue #5579, when ``rename`` fails because of different filesystems of source and destination.

### CODE REVIEW

File: lib/pure/os.nim [hunk 1]
```diff
@@ -666,6 +652,30 @@ proc removeFile*(file: string) {.rtl, extern: "nos$1", tags: [WriteDirEffect].}
...
+      if err == EXDEV.OSErrorCode:
+        # Fallback to copy & del
+        copyFile(source, dest)
+        try:
+          removeFile(source)
+        except:
+          try: removeFile(dest) except: discard
+          raise
+      else:
+        raiseOSError(osLastError(), $strerror(errno))
```
Comment (Araq) [inline lib/pure/os.nim]: Use ``err``  here.

File: lib/pure/os.nim [hunk 2]
```diff
@@ -666,6 +652,30 @@ proc removeFile*(file: string) {.rtl, extern: "nos$1", tags: [WriteDirEffect].}
...
+  else:
+    if c_rename(source, dest) != 0'i32:
+      let err = osLastError()
+      if err == EXDEV.OSErrorCode:
+        # Fallback to copy & del
+        copyFile(source, dest)
+        try:
+          removeFile(source)
+        except:
+          try: removeFile(dest) except: discard
```
Comment (Araq) [inline lib/pure/os.nim]: I don't like these 2 try statements here. Ideally it should be something like:

```nim
if copyFile(source, dest): 
  removeFile(source)
else:
  discard tryRemoveFile(dest)
```

Comment (vegansk) [inline lib/pure/os.nim]: Removed inner try. But if I'll remove another one, the exception information will be lost

File: lib/pure/os.nim (hunk 1 above)
Comment (vegansk) [inline lib/pure/os.nim]: Done

**Comment by vegansk** [comment]
@Araq, there was the bug in the old version of removeFile, see the code below. I fixed it, but it seems that is is a breaking change for windows. Should I continue to fix windows build, or should I just restore this bug and save backward compatibility?
```nim
proc removeFile*(file: string) {.rtl, extern: "nos$1", tags: [WriteDirEffect].} =
  when defined(Windows):
    when useWinUnicode:
      let f = newWideCString(file)
    else:
      let f = file
    if deleteFile(f) == 0:
      # It will nether throw the error if getLastError() != ERROR_ACCESS_DENIED:
      if getLastError() == ERROR_ACCESS_DENIED:
        if setFileAttributes(f, FILE_ATTRIBUTE_NORMAL) == 0:
          raiseOSError(osLastError())
        if deleteFile(f) == 0:
          raiseOSError(osLastError())
  else:
    if c_remove(file) != 0'i32 and errno != ENOENT:
      raiseOSError(osLastError(), $strerror(errno))
```

**Comment by Araq** [comment]
It fails with:

```
Traceback (most recent call last)
niminst.nim(748)         niminst
niminst.nim(540)         srcdist
niminst.nim(493)         deduplicateFiles
os.nim(662)              removeFile
os.nim(154)              raiseOSError
Error: unhandled exception: The system cannot find the path specified.
 [OSError]
Error: execution of an external program failed: 'C:\projects\nim\tools\nim
```

And since the logic in niminst is quite involved, let's assume that's a bug in niminst. So ... we need to fix niminst.

**Comment by Araq** [comment]
Thinking about it... niminst also runs on Linux without issues, so your "fix" of ``removeFile``  seems to blame. Are you aware it shouldn't fail for "file not found"? That's what its docs say iirc.

**Comment by vegansk** [comment]
Updated, now it handles ERROR_FILE_NOT_FOUND and ERROR_PATH_NOT_FOUND

**Pull request merged**